src/Stackage/Package/Git/Object.hs

```diff
@@ -0,0 +1,147 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+module Stackage.Package.Git.Object
+  ( GitObject(..)
+  , makeGitFile
+  , makeGitCommit
+  , makeGitTag
+  , repoWriteObject
+  ) where
+
+import ClassyPrelude.Conduit
+import qualified Codec.Compression.Zlib as Zlib
+import Data.ByteArray (convert)
+import Crypto.Hash (SHA1(..), Digest)
+import Crypto.Hash.Conduit (sinkHash)
+import qualified Data.ByteString as S
+import qualified Data.ByteString.Char8 as S8
+import qualified Data.ByteString.Lazy as L
+import Data.Conduit.Zlib (compress, defaultWindowBits)
+import qualified Data.Git.Types as G
+import Data.Git.Storage
+import Data.Git.Storage.Object
+import Data.Git.Ref
+import Data.Git.Storage.Loose
+import System.Directory
+import System.FilePath
+
+
+import Stackage.Package.Git.Types
+
+
+-- | An an object that can be written to git store.
+data GitObject
+  = Blob GitFile
+  | Tree GitTree
+  | Commit GitCommit
+  | Tag GitTag
+
+
+-- | Marshalls a bytestring into a git blob object, computes it's SHA1 and
+-- compresses it and returnes a `GitFile` that is of type `NonExecFile`.
+makeGitFile
+  :: (MonadThrow m, PrimMonad base, MonadBase base m)
+  => LByteString -- ^ Content of the blob.
+  -> Word64 -- ^ Size of the content.
```

Comment:
I thought about that, and it was an optimization I chose to do. This function is used in two places:
`Git.Repository.repoWriteFile` and `IndexConduit.makeGitFile`. The latter place has size of the `LByteString` available from the tar file (and if it is incorrect in tar, then there is a bigger problem), so using that information we can avoid `O(n/c)` calls to `length`, considering that this is called about 70K time during each update I though it might be worthwhile optimization, although at a slight cost of safety. In this approach there is another possible safety concern, that I probably should handle separately, namely the size from tar is `Int64`, while we expect `Word64`.
In any case, I don't think there is much performance gained from this, so having explained my reasoning, would you like me to go the safe route and just use `length lbs`?